### ISSUE TITLE
Android 10 audio forwarding

### ISSUE THREAD

**Issue by fenste** [Genymobile/scrcpy#933]
Are you considering adding audio forwarding for Android 10 when they added [https://developer.android.com/guide/topics/media/playback-capture](url) ?

**Comment by rom1v** [comment]
Absolutely: https://github.com/Genymobile/scrcpy/issues/14#issuecomment-523071057

**Comment by AndroidDeveloperLB** [comment]
Will it be possible officially, then? Even for Windows?

**Comment by rom1v** [comment]
If it works, yes. That would be great.

The problem is finding time to work on it. :disappointed:

**Comment by AndroidDeveloperLB** [comment]
Why don't Google offer an official API for this? They have an adb command for recording, but not including audio...
How do you use adb for the mirroring anyway?

**Comment by rom1v** [comment]
> How do you use adb for the mirroring anyway?

I explained [here](https://blog.rom1v.com/2018/03/introducing-scrcpy/) and [here](https://github.com/Genymobile/scrcpy/blob/master/DEVELOP.md).

**Comment by AndroidDeveloperLB** [comment]
Wait, you somehow run code on the device, without installing anything?
How is it even possible?
I don't get it...

**Comment by LBensman** [comment]
> Wait, you somehow run code on the device, without installing anything?
> How is it even possible?
> I don't get it...

Installing an "app" via an app store is not the only way to get something running on an Android device.  You can push files to device via `adb push` and you can execute those files using `app_process` command on the device itself (usually via `adb shell` command executed on the host).

**Comment by AndroidDeveloperLB** [comment]
I said "without installing anything". 
You can push a file and run something in it, without installing anything?

**Comment by LBensman** [comment]
> I said "without installing anything".
> You can push a file and run something in it, without installing anything?

Sure, why not?  There are limitations, I think, as to what a process can do that is launched that way, so you can still be limited in what APIs you can call from such a process, but at least for grabbing screen surface and video encoding, it has sufficient means to do so.

**Comment by AndroidDeveloperLB** [comment]
This is incredible. What are the limitations, then?
It can get all that's on the screen. That's a lot, no? 
How could it get the screen content? I thought that maybe you will be restricted to pure Java stuff, but this is even beyond what Android's framework offers...

**Comment by LBensman** [comment]
I'm not exactly sure of the limitations or what would constitute being a lot or a little, but take a look at [`SurfaceControl` wrapper](https://github.com/Genymobile/scrcpy/blob/bdd05b4a16a2b0fc15d2ecf8f832afb9670854ee/server/src/main/java/com/genymobile/scrcpy/wrappers/SurfaceControl.java) to see what it calls to get the screen data, as well how it is [wired as input to encoder](https://github.com/Genymobile/scrcpy/blob/bdd05b4a16a2b0fc15d2ecf8f832afb9670854ee/server/src/main/java/com/genymobile/scrcpy/ScreenEncoder.java#L70).

**Comment by AndroidDeveloperLB** [comment]
Amazing. Seems you used a lot of reflection though. 
How did you find about this? How did you know it's even possible?

**Comment by LBensman** [comment]
Oh, I'm not the author of it - you'd have to ask @rom1v on how he did the research and the level of experimentation that it took :)

**Comment by AndroidDeveloperLB** [comment]
OK just know that I had this expression like you have on your avatar...
haha

**Comment by rom1v** [comment]
https://developer.android.com/guide/topics/media/playback-capture

> The app must have the `RECORD_AUDIO` permission.

The [shell](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/packages/Shell/AndroidManifest.xml) does not have this permission :cry: 

So it will require to install an apk. :disappointed:

**Comment by agent47nh** [comment]
So why not keep an _optional APK_ just for audio?

**Comment by arex0** [comment]
In my phone android 10 and desktop debain testing, [https://android.stackexchange.com/questions/58026/how-to-stream-audio-from-android-to-pulseaudio-server](https://android.stackexchange.com/questions/58026/how-to-stream-audio-from-android-to-pulseaudio-server) working

**Comment by AndroidDeveloperLB** [comment]
I don't mind having an additional APK for this, as long as it doesn't ruin other things.
What happens if multiple apps try to use the mic?
Maybe you could add an option in the window of scrcpy to turn it on/off?

**Comment by rom1v** [comment]
- https://blog.rom1v.com/2020/06/audio-forwarding-on-android-10/
 - https://github.com/rom1v/sndcpy

**Comment by AndroidDeveloperLB** [comment]
@rom1v Either I don't get how to make it work, or it doesn't work.

I tried this:
1. Install the apk from there.
2. Run scrcpy.
3. Run sndcpy.
4. Grant permission on the device to capture video.
5. Press Enter on the console window.
6. Play some video with audio on YouTube app on the device.

Nothing occurs

**Comment by rom1v** [comment]
> It shows a notification of "sndcpy" with "Waiting for connection..."

Did you press "Enter" in the sndcpy terminal?

What's the output in your terminal?

**Comment by AndroidDeveloperLB** [comment]
Yes, that's step that I wrote about.
BTW, Pixel 4 with Android 10.

There is no output on the terminal , because it was closed when I pressed Enter...

**Comment by rom1v** [comment]
OK, do you have VLC installed?

Try to run it from `cmd.exe` instead of double-clicking on the `.bat`.

**Comment by AndroidDeveloperLB** [comment]
VLC is installed on the PC, but do I need to start it?
Other than seeing that it should be installed, I don't see any other mention of it.

**Comment by rom1v** [comment]
> Other than seeing that it should be installed, I don't see any other mention of it.

Read `sndcpy.bat` :wink:

**Comment by AndroidDeveloperLB** [comment]
That's step 3. Already done this.

And I tried to run VLC on my own too.

**Comment by AndroidDeveloperLB** [comment]
Wait, you wrote "Read " ?

**Comment by rom1v** [comment]
No, I meant [read](https://github.com/rom1v/sndcpy/blob/master/sndcpy.bat). You will understand how VLC is used :)

**Comment by AndroidDeveloperLB** [comment]
Read? But it doesn't say anything when running it, about VLC.
If you mean read in source code,  why read in source code instead of instructions of how to use the new tool?
Can you please tell me what to do?

**Comment by rom1v** [comment]
> But it doesn't say anything when running it, about VLC.

What is the output of the terminal when you run `sndcpy`? (if you run it manually from `cmd.exe` so that the terminal does not close automatically)

> If you mean read in source code, why read in source code instead of instructions of how to use the new tool?

As a user, you don't really need to know. Running `sndcpy` should be sufficient. That's why you "don't see any other mention of it" (except in the script source code).

**Comment by AndroidDeveloperLB** [comment]
As I wrote, when pressing Enter, it closes.
If you mean that I should run sndcpy from console and then press Enter, this is what I see:

```
Waiting for device...
Performing Streamed Install
Success
Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=com.rom1v.sndcpy/.MainActivity }
Press Enter once audio capture is authorized on the device to start playing...
Playing audio...
The system cannot find the path specified.
```

The weird thing is that when I tried on another PC (a laptop actually), it worked fine.
The other PC has Intel, and mine has AMD. Is it related?

**Comment by rom1v** [comment]
>     The system cannot find the path specified.

This is the important message.

Is your vlc installed in `C:\Program Files\VideoLAN\VLC\vlc.exe`? If not, pass an environment variable: https://github.com/rom1v/sndcpy/blob/master/README.md#environment-variables

**Comment by AndroidDeveloperLB** [comment]
For some reason it decided to be installed on "C:\Program Files (x86)\VideoLAN\VLC\vlc.exe" instead.

I decided to uninstall it, and re-install it, because maybe it installed as 32 bit or something. Now it's on the path you wrote.
But now it doesn't close the console window, and yet I still can't hear from the PC.

This is what I see on the console this time:

```
Waiting for device...
Performing Streamed Install
Success
Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=com.rom1v.sndcpy/.MainActivity }
Press Enter once audio capture is authorized on the device to start playing...
Playing audio...
```

There is a bug on the Android app though: it keeps listening. Never stops on its own. Even if I disconnect from the USB.
Is it by design?
I'd prefer if scrcpy could be run with all of these using a single shortcut that I just launch. Maybe via a parameter to it "-forward-audio" or something.

**Comment by rom1v** [comment]
> But now it doesn't close the console window

:+1:

> and yet I still can't hear from the PC.

What app do you test on your device? Try youtube or VLC android for example.

I think we could continue on https://github.com/rom1v/sndcpy/issues/1#issuecomment-641864289.

**Comment by AndroidDeveloperLB** [comment]
I tried on YouTube.

**Comment by rom1v** [comment]
> The weird thing is that when I tried on another PC (a laptop actually), it worked fine.

The same device works on another laptop?

> yet I still can't hear from the PC.

But on this laptop, no sound?

**Comment by AndroidDeveloperLB** [comment]
Please let's continue there.

**Comment by AndroidDeveloperLB** [comment]
OK I tested again today.
This time on Android R.
I've noticed something weird:
It plays on both the PC and device, so I hear twice, with a small delay between them
On the device the volume can be 0 though, so I won't hear anything from it, while hearing from PC.

**Comment by rom1v** [comment]
:+1:

> It plays on both the PC and device

Yes, that's the expected behavior (the playback capture API works this way).

**Comment by AndroidDeveloperLB** [comment]
Shouldn't it at least set the volume to minimal?
It's weird to hear the same thing with a delay on both PC and device...
Is there perhaps a parameter for it, so that it will try to restore volume when it's done, or something?

**Comment by huanhor** [comment]
> Shouldn't it at least set the volume to minimal?
> It's weird to hear the same thing with a delay on both PC and device...
> Is there perhaps a parameter for it, so that it will try to restore volume when it's done, or something?

Hey friend,
I know your problem and i solve it. You try to install the version of VLC are the Windows 64bit and try again

**Comment by yume-chan** [comment]
Since this one is still open...

I found https://infosecwriteups.com/exploiting-android-zygote-injection-cve-2024-31317-d83f69265088, which gives system permission (uid 1000).

This is not patched for Android 10 and 11, so Scrcpy (and attackers) can still use it.

I tested it in my own client, with some minor tweaks to the server to remove version checks:

<details>
<summary>patch</summary>

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/FakeContext.java b/server/src/main/java/com/genymobile/scrcpy/FakeContext.java
index 22fc6d49..4e1ae2ae 100644
--- a/server/src/main/java/com/genymobile/scrcpy/FakeContext.java
+++ b/server/src/main/java/com/genymobile/scrcpy/FakeContext.java
@@ -10,10 +10,11 @@ import android.content.ContextWrapper;
 import android.content.IContentProvider;
 import android.os.Binder;
 import android.os.Process;
+import android.system.Os;
 
 public final class FakeContext extends ContextWrapper {
 
-    public static final String PACKAGE_NAME = "com.android.shell";
+    public static final String PACKAGE_NAME = Os.getuid() == 2000 ? "com.android.shell" : "";
     public static final int ROOT_UID = 0; // Like android.os.Process.ROOT_UID, but before API 29
 
     private static final FakeContext INSTANCE = new FakeContext();
diff --git a/server/src/main/java/com/genymobile/scrcpy/Server.java b/server/src/main/java/com/genymobile/scrcpy/Server.java
index 09cfd6cf..afaa0bb6 100644
--- a/server/src/main/java/com/genymobile/scrcpy/Server.java
+++ b/server/src/main/java/com/genymobile/scrcpy/Server.java
@@ -25,6 +25,7 @@ import com.genymobile.scrcpy.video.SurfaceEncoder;
 import com.genymobile.scrcpy.video.VideoSource;
 
 import android.os.Build;
+import android.system.Os;
 
 import java.io.File;
 import java.io.IOException;
@@ -75,7 +76,12 @@ public final class Server {
     }
 
     private static void scrcpy(Options options) throws IOException, ConfigurationException {
-        if (Build.VERSION.SDK_INT < AndroidVersions.API_31_ANDROID_12 && options.getVideoSource() == VideoSource.CAMERA) {
+        Ln.i("Current uid is: "+ Os.getuid());
+
+        if (Build.VERSION.SDK_INT < AndroidVersions.API_31_ANDROID_12 &&
+                options.getVideoSource() == VideoSource.CAMERA &&
+                Os.getuid() != 0 &&
+                Os.getuid() != 1000) {
             Ln.e("Camera mirroring is not supported before Android 12");
             throw new ConfigurationException("Camera mirroring is not supported");
         }
diff --git a/server/src/main/java/com/genymobile/scrcpy/audio/AudioDirectCapture.java b/server/src/main/java/com/genymobile/scrcpy/audio/AudioDirectCapture.java
index 5c859738..14d9c91e 100644
--- a/server/src/main/java/com/genymobile/scrcpy/audio/AudioDirectCapture.java
+++ b/server/src/main/java/com/genymobile/scrcpy/audio/AudioDirectCapture.java
@@ -15,6 +15,7 @@ import android.media.MediaCodec;
 import android.media.MediaRecorder;
 import android.os.Build;
 import android.os.SystemClock;
+import android.system.Os;
 
 import java.nio.ByteBuffer;
 
@@ -126,7 +127,9 @@ public class AudioDirectCapture implements AudioCapture {
 
     @Override
     public void start() throws AudioCaptureException {
-        if (Build.VERSION.SDK_INT == AndroidVersions.API_30_ANDROID_11) {
+        if (Build.VERSION.SDK_INT == AndroidVersions.API_30_ANDROID_11 &&
+                Os.getuid() != 0 &&
+                Os.getuid() != 1000) {
             startWorkaroundAndroid11();
             try {
                 tryStartRecording(5, 100);
```

</details>

On Samsung S9 running Android 10, audio and camera both work.

On Android 11, new SELinux rules no longer allow uid 1000 to read `/data/local/tmp` folder, so we need another method to start Scrcpy.
